db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/ExecutePicked.java

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright (C) Gustav Karlsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.kagkarlsson.scheduler;
+
+import com.github.kagkarlsson.scheduler.stats.StatsRegistry;
+import com.github.kagkarlsson.scheduler.task.CompletionHandler;
+import com.github.kagkarlsson.scheduler.task.Execution;
+import com.github.kagkarlsson.scheduler.task.ExecutionComplete;
+import com.github.kagkarlsson.scheduler.task.ExecutionContext;
+import com.github.kagkarlsson.scheduler.task.ExecutionOperations;
+import com.github.kagkarlsson.scheduler.task.FailureHandler;
+import com.github.kagkarlsson.scheduler.task.Task;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.util.Optional;
+import java.util.UUID;
+
+@SuppressWarnings("rawtypes")
+class ExecutePicked implements Runnable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ExecutePicked.class);
+    private final Scheduler scheduler;
+    private final Execution pickedExecution;
+
+    public ExecutePicked(Scheduler scheduler, Execution pickedExecution) {
```

Comment:
While working on this code, I faced some challenges in testing/following the code and updating due to the mutual dependency of Scheduler on the Execution and vice versa.

Could we have a more tree-like structure in a way that ExecutePicked gets "trigger"/"handler" interface instead of the scheduler itself?

Reply:
Yeah, this a bit of technical debt I have not fixed yet. I do not like passing the full `Scheduler` here. It should be changed to not reference `Scheduler`.. haven't taken the time to figure out a good solution yet..